docs/architecture.md

```diff
@@ -15,139 +15,207 @@ Prometheus instances scrape samples from various targets and then push them to C
+#### HA Tracker
 
-#### Hashing
+The distributor features an **HA Tracker**. When enabled, the distributor deduplicates incoming samples from redundant Prometheus servers. This allows you to have multiple HA replicas of the same Prometheus servers, writing the same series to Cortex and then deduplicate these series in the Cortex distributor.
+
+The HA Tracker deduplicates incoming samples based on a cluster and replica label. The cluster label uniquely identifies the cluster of redundant Prometheus servers for a given tenant, while the replica label uniquely identify the replica within the Prometheus servers cluster. Incoming samples are considered duplicated - and thus skipped - if received by any replica which is not the current primary within a cluster.
 
-Distributors use consistent hashing, in conjunction with the (configurable) replication factor, to determine *which* instances of the ingester service receive each sample.
+The HA Tracker requires a key-value (KV) store to coordinate which is the currently elected primary replica from which the distributor accepts the samples. Samples with one or neither of the replica and cluster labels are accepted by default and never deduplicated.
 
-The hash itself is based on one of two schemes:
+The supported KV stores for the HA tracker are:
 
-1. The metric name and tenant ID
-2. All the series labels and tenant ID
+* [Consul](https://www.consul.io)
+* [Etcd](https://etcd.io)
+
+For more information, please check out the [Config for sending HA pairs data to Cortex](guides/ha-pair-handling.md) documentation.
+
+#### Hashing
 
-The trade-off associated with the latter is that writes are more balanced but they must involve every ingester in each query.
+Distributors use consistent hashing, in conjunction with a configurable replication factor, to determine which instances of the ingester service should receive a given series.
 
-> This hashing scheme was originally chosen to reduce the number of required ingesters on the query path. The trade-off, however, is that the write load on the ingesters is less even.
+Cortex supports two hashing strategies:
+
+1. Hash the metric name and tenant ID (default)
+2. Hash the metric name, labels and tenant ID (recommended, enabled with `-distributor.shard-by-all-labels=true`)
+
+The trade-off associated with the latter is that writes are more balanced across ingesters but each query needs to talk to any ingester since a metric could be spread across any ingester given different label sets. The first hashing strategy (default) was originally chosen to reduce the number of required ingesters on the query path, but running Cortex in production showed that an evenly distributed write load on the ingesters is preferable at the cost of always querying all ingesters.
```

Comment:
I agree on being clear about the reason.

> increases the likelihood of a query needing

It's not that it increases the likelihood, but we literally always query all ingesters. I've cherry picked your change and changed that part.

Feel free to further comment if you have any other feedback.

Reply:
Thank you for clearing that up! Looks good 👌 

The diff still shows the previous paragraph, not sure if this is indented or not but was worth mentioning.

Reply:
one tiny nit: Is that we use the word "spread" followed by "any" which suggests a division by 1. My suggestion is:

> since a metric could be spread across multiple ingesters

If my understanding of how this works is correct.

Reply:
> The diff still shows the previous paragraph, not sure if this is indented or not but was worth mentioning.

Ops, fixed. Thanks!

> since a metric could be spread across multiple ingesters

Way better.